language-support/ts/daml-ledger/index.ts

```diff
@@ -645,7 +719,98 @@ class Ledger {
+    return this.streamSubmit("streamFetchByKey", template, 'v1/stream/fetch', request, reconnectRequest, null, change);
+  }
+
+  /**
+   * @internal
+   *
+   * Returns the same API as [[streamSubmit]] but does not, in fact, establish
+   * any socket connection. Instead, this is a stream that always has the given
+   * value as its accumulated state.
+   */
+  private constantStream<T extends object, K, I extends string, V>(
+    value: V,
+  ): Stream<T, K, I, V> {
+    function on(t: 'live', l: (v: V) => void): void;
+    function on(t: 'change', l: (v: V, events: readonly Event<T, K, I>[]) => void): void;
+    function on(t: 'close', l: (e: StreamCloseEvent) => void): void;
+    // eslint-disable-next-line @typescript-eslint/no-explicit-any
+    function on(type: any, listener: any): any {
+        if (type === 'live') {
+          listener(value);
+        }
+        if (type === 'change') {
+          listener(value, []);
+        }
+    }
+    function off(t: 'live', l: (v: V) => void): void;
+    function off(t: 'change', l: (v: V, events: readonly Event<T, K, I>[]) => void): void;
+    function off(t: 'close', l: (e: StreamCloseEvent) => void): void;
+    // eslint-disable-next-line @typescript-eslint/no-explicit-any, @typescript-eslint/no-unused-vars, @typescript-eslint/no-empty-function
+    function off(_t: any, _l: any): any {}
+    // eslint-disable-next-line @typescript-eslint/no-empty-function
+    return { on, off, close: (): void => {} };
+  }
+
+  /**
+   * Retrieve a consolidated stream of events for a list of keys and a single
+   * template.
+   *
+   * The accumulated state is an array of the same length as the given list of
+   * keys, with positional correspondence. Each element in the array represents
+   * the current contract for the given key, or is explicitly null if there is
+   * currently no active contract matching that key.
+   *
+   * Note: the given `key` objects will be compared for (deep) equality with
+   * the values returned by the API. As such, they have to be given in the
+   * "output" format of the API, including the values of
+   * `encodeDecimalAsString` and `encodeInt64AsString`. See [the JSON API docs
+   * for details](https://docs.daml.com/json-api/lf-value-specification.html).
+   *
+   * @typeparam T The contract template type.
+   * @typeparam K The contract key type.
+   * @typeparam I The contract id type.
+   */
+  streamFetchByKeys<T extends object, K, I extends string>(
+    template: Template<T, K, I>,
+    keys: K[],
+  ): Stream<T, K, I, (CreateEvent<T, K, I> | null)[]> {
+    // We support zero-length key so clients can more easily manage a dynamic
+    // list, without having to special-case 0-length on their side.
+    if (keys.length == 0) {
+      return this.constantStream([]);
+    }
+    const lastContractIds: (ContractId<T> | null)[] = Array(keys.length).fill(null);
+    const keysCopy = deepCopy(keys);
```

Comment:
> Makes sense here. In general, I’m a bit hesitant to make that guarantee since we don’t test this anywhere and I wouldn’t be surprised if in various places, we do make assumptions about the format.

I'm not sure I understand what "guarantee" this is talking about. I have added to the documentation that users need to provide the key in the documented "output" format of the JSON API, but this is more like a requirement than a guarantee.

> Why do we need to create a copy here? Just to protect against the caller mutating the keys object? If so what makes this parameter special? Any object is passed by reference and can be mutated so we would need to do this for a lot of stuff and I’m not sure if that’s particularly idiomatic for JS.

As far as I'm aware it is idiomatic (and, more importantly, correct), though it's been a while since I last participated in the JS community.

What makes this parameter special, and thus worth copying, is that we keep it around for later reference through asynchronous callbacks. With JS not having threads, composite objects captured by closures are the only case where you need to be mindful of mutability, which is a bit less burdensome (if not less error-prone) than all arguments all the time. At least in this PR and barring an error on my side, this is the only closed-over composite.

Reply:
> I'm not sure I understand what "guarantee" this is talking about.

I was coming from your statement that we advertise this as JS bindings and should support inputs even if they do not pass the TS typechecker unless documented differently (this being the guarantee). That just seems hard to keep up without testing this or even knowing where it works by accident (and what exactly works) and where it fails. The documentation on the new methods and the behavior there is perfectly fine and I fully agree with that.

Your explanation on this parameter being special makes sense so happy to keep it.